src/coreclr/jit/importer.cpp

```diff
@@ -840,109 +872,92 @@ void Compiler::impPopCallArgs(unsigned count, CORINFO_SIG_INFO* sig, GenTreeCall
-                {
-                    // Make sure that all valuetypes (including enums) that we push are loaded.
-                    // This is to guarantee that if a GC is triggered from the prestub of this methods,
-                    // all valuetypes in the method signature are already loaded.
-                    // We need to be able to find the size of the valuetypes, but we cannot
-                    // do a class-load from within GC.
-                    info.compCompHnd->classMustBeLoadedBeforeCodeIsRun(argRealClass);
-                }
-            }
+            argNode = impImplicitIorI4Cast(argNode, jitSigType);
+        }
 
-            const var_types nodeArgType = arg.GetEarlyNode()->TypeGet();
-            if (!varTypeIsStruct(jitSigType) && genTypeSize(nodeArgType) != genTypeSize(jitSigType))
+        if (corType != CORINFO_TYPE_CLASS && corType != CORINFO_TYPE_BYREF && corType != CORINFO_TYPE_PTR &&
+            corType != CORINFO_TYPE_VAR)
+        {
+            CORINFO_CLASS_HANDLE argRealClass = info.compCompHnd->getArgClass(sig, sigArg);
+            if (argRealClass != nullptr)
             {
-                assert(!varTypeIsStruct(nodeArgType));
-                // Some ABI require precise size information for call arguments less than target pointer size,
-                // for example arm64 OSX. Create a special node to keep this information until morph
-                // consumes it into `CallArgs`.
-                GenTree* putArgType = gtNewOperNode(GT_PUTARG_TYPE, jitSigType, arg.GetEarlyNode());
-                arg.SetEarlyNode(putArgType);
+                // Make sure that all valuetypes (including enums) that we push are loaded.
+                // This is to guarantee that if a GC is triggered from the prestub of this methods,
+                // all valuetypes in the method signature are already loaded.
+                // We need to be able to find the size of the valuetypes, but we cannot
+                // do a class-load from within GC.
+                info.compCompHnd->classMustBeLoadedBeforeCodeIsRun(argRealClass);
             }
+        }
 
-            sigArgs = info.compCompHnd->getArgNext(sigArgs);
+        const var_types nodeArgType = argNode->TypeGet();
+        if (!varTypeIsStruct(jitSigType) && genTypeSize(nodeArgType) != genTypeSize(jitSigType))
+        {
+            assert(!varTypeIsStruct(nodeArgType));
+            // Some ABI require precise size information for call arguments less than target pointer size,
+            // for example arm64 OSX. Create a special node to keep this information until morph
+            // consumes it into `CallArgs`.
+            argNode = gtNewOperNode(GT_PUTARG_TYPE, jitSigType, argNode);
+        }
 
-            index++;
+        if (lastArg == nullptr)
+        {
+            lastArg = call->gtArgs.PushFront(this, argNode);
         }
+        else
+        {
+            lastArg = call->gtArgs.InsertAfter(this, lastArg, argNode);
```

Comment:
One thing I'm not sure if I I need to worry about here is the fact that in DEBUG builds, `InsertAfter` does a `O(n)` loop over args to verify that `lastArg` is contained in `CallArgs`. It essentially makes this entire loop `O(n^2)` for debug/checked builds. We can support up to approximately 2^16 args so that would potentially become quite expensive for a pathological case. Should I introduce `InsertAfterUnchecked` to use here?

Reply:
Looks like we have a context in coreclr_tests with a call with 8192 args... I'll introduce an unchecked variant.